File: days/day_20/main.py
import math
import time
import os


class Piece:
	def __init__(self, num, grid):
		self.num = num
		self.grid = grid
		self.d = len(grid)

	def top(self):
		return self.grid[0]

	def bot(self):
		return self.grid[self.d-1]

	def left(self):
		return [row[0] for row in self.grid]

	def right(self):
		return [row[self.d-1] for row in self.grid]

	# flip along vertical axis
	def flip(self):
		for row in self.grid:
			row.reverse()

	# rotate 90 degrees clockwise
	def rotate(self):
		self.grid = [list(row) for row in zip(*reversed(self.grid))]

	def __str__(self):
		return str(self.grid)

	def __repr__(self):
		return str(self.num)


def solve(board, pieces, n, row, col):
	if col + n*row < n*n: 
		if col == n:
			row += 1
			col = 0
		
		for piece in pieces:
			board[row][col] = piece
			others = pieces.copy()
			others.remove(piece)

			# check all orientations
			for i in range(4):
				board[row][col].rotate()
				if check(board, row, col):
					if solve(board, others, n, row, col+1):
						return True
			
			board[row][col].flip()

			for i in range(4):
				board[row][col].rotate()
				if check(board, row, col):
					if solve(board, others, n, row, col+1):
						return True

			# backtrack
			board[row][col] = 0
		
		# no solution
		return False

	else:
		return True


def check(board, row, col):
	if row > 0:
		if board[row][col].top() != board[row-1][col].bot():
			return False
	if col > 0:
		if board[row][col].left() != board[row][col-1].right():
			return False
	return True


def get_image(board):
	img = []
	line = []
	n = len(board)
	d = board[0][0].d
	i = 0

	while i < n*d:
		if i%d not in [0,d-1]:  
			for piece in board[i//d]:
				line.extend(piece.grid[i%d][1:-1])

			img.append(''.join(line))
			line = []

		i+=1

	return img


def flipped(grid):
	return [list(reversed(row)) for row in grid]


def rotated(grid):
	return [list(row) for row in zip(*reversed(grid))]


def get_orientations(grid):
	grids = []

	grids.append(grid)

	for i in range(3):
		grids.append(rotated(grids[-1]))
	
	grids.append(flipped(grids[0]))

	for i in range(3):
		grids.append(rotated(grids[-1]))

	return grids


def part_one(pieces):
	n = int(math.sqrt(len(pieces)))
	board = [[0 for i in range(n)] for j in range(n)]

	solve(board, pieces, n, 0, 0)

	return (int(board[0][0].num)
					* int(board[n-1][0].num)
					* int(board[0][n-1].num)
					* int(board[n-1][n-1].num),
					board)


def part_two(board):
	img = [list(line) for line in get_image(board)]

	monsters = get_orientations(['                  # ',
															 '#    ##    ##    ###',
															 ' #  #  #  #  #  #   '])

	for monster in monsters:
		h = len(monster)
		w = len(monster[0])
		
		for x in range(len(img)-w):
			for y in range(len(img)-h):
				
				found = True
				
				for i in range(w):
					if not found:
						break
					for j in range(h):
						if monster[j][i] not in [' ', img[y+j][x+i]]:
							if img[y+j][x+i] != 'O':
								found = False
				
				# mark monster
				if found:
					for i in range(w):
						for j in range(h):
							if monster[j][i] != ' ':
								img[y+j][x+i] = 'O'

	count = 0
	for line in img:
		for x in line:
			if x == '#':
				count += 1

	return count


def main():
	start_time = time.time()

	with open(os.path.dirname(__file__) + '/input.txt', 'r') as data:
		lines = [line.strip() for line in data.readlines()]

		pieces = []

		i = 0
		num = -1
		grid = []
		
		while i < len(lines):
			line = lines[i]

			if line == '':
				pieces.append(Piece(num, grid))
				num = -1
				grid = []
			else:
				if num == -1:
					num = int(line[5:].strip(':'))
				else:
					grid.append(list(line))

			i += 1
		
		pieces.append(Piece(num, grid))

		part_one_ans, board = part_one(pieces)
		part_two_ans = part_two(board)

		print('day 20  ({:,.3f}s)'.format(time.time()-start_time))
		print('  part 1: {}'.format(part_one_ans))
		print('  part 2: {}'.format(part_two_ans))


if __name__ == "__main__":
		main()

File: days/day_20/main.js
const fs = require('fs');
const readline = require('readline');

async function loadInput() {
  const rl = readline.createInterface({
    input: fs.createReadStream(__dirname + '/input.txt')
  });

  const lines = [];
  for await (const line of rl) {
    lines.push(line);
  }
  return lines;
}

class Piece {
  constructor(num, grid) {
    this.num = num;
    this.grid = grid;
    this.d = grid.length;
  }

  top() {
    return this.grid[0];
  }

  bot() {
    return this.grid[this.d-1];
  }

  left() {
    return this.grid.map(row => row[0]);
  }

  right() {
    return this.grid.map(row => row[this.d-1]);
  }
  
  flip() {
    for (const row of this.grid) {
      row.reverse();
    }
  }

  rotate() {
    this.grid = this.grid[0].map((_, i) => this.grid.map(row => row[this.d-1-i]));
  }


  toString() {
    return String(this.num);
  }
}


function solve(board, pieces, n, row, col) {
  if ((col + n*row) < n*n) { 
    if (col == n) {
      row += 1;
      col = 0;
    }
    
    for (const piece of pieces) {
      board[row][col] = piece;
      const others = [...pieces].filter(other => other.num != piece.num);

      // check all orientations
      for (let i = 0; i < 4; i++) {
        board[row][col].rotate();
        if (check(board, row, col)) {
          if (solve(board, others, n, row, col+1)) {
            return true;
          }
        }
      }  
      
      board[row][col].flip();

      for (let i = 0; i < 4; i++) {
        board[row][col].rotate();
        if (check(board, row, col)) {
          if (solve(board, others, n, row, col+1)) {
            return true;
          }
        }
      }

      // backtrack
      board[row][col] = 0;
    }
    
    //no solution
    return false;
  }
  else {
    return true;
  }
}

function arrayEquals(a, b) {
  for (let i = 0; i < a.length; ++i) {
    if (a[i] !== b[i]) {
      return false;
    }
  }
  return true;
}

function check(board, row, col) {
  if (row > 0) {
    if (!arrayEquals(board[row][col].top(), board[row-1][col].bot())) {
      return false;
    }
  }
  if (col > 0) {
    if (!arrayEquals(board[row][col].left(), board[row][col-1].right())) {
      return false;
    }
  }
  return true;
}

function getImage(board) {
  const img = [];
  const n = board.length;
  const d = board[0][0].d;

  let line = [];
  let i = 0;

  while (i < n*d) {
    if (![0,d-1].includes(i%d)) {  
      for (const piece of board[Math.floor(i/d)]) {
        line.push(...piece.grid[i%d].slice(1, -1));
      }

      img.push([...line]);
      line = [];
    }
    i += 1;
  }

  return img;
}

function flipped(grid) {
  return grid.map(row => row.slice().reverse());
}

function rotated(grid) {
  const d = grid[0].length;
  return grid[0].map((_, i) => grid.map(row => row[d-1-i]));
}

function getOrientations(grid) {
  const grids = [];

  grids.push(grid);

  for (let i = 0; i < 3; i++) {
    grids.push(rotated(grids[i]));
  }
  
  grids.push(flipped(grids[0]));

  for (let i = 0; i < 3; i++) {
    grids.push(rotated(grids[i+4]));
  }

  return grids;
}

function partOne(pieces) {
  const n = Math.sqrt(pieces.length);
  const board = [...Array(n)].map(row => Array(n).fill(0));

  solve(board, pieces, n, 0, 0);

  return [Number(board[0][0].num)
          * Number(board[n-1][0].num)
          * Number(board[0][n-1].num)
          * Number(board[n-1][n-1].num),
          board];
}

function partTwo(board) {
  const img = getImage(board);
  const monsters = getOrientations(['                  # '.split(''),
                                    '#    ##    ##    ###'.split(''),
                                    ' #  #  #  #  #  #   '.split('')]);

  for (const monster of monsters) {
    h = monster.length;
    w = monster[0].length;
    
    for (let x = 0; x < img.length - w; x++) {
      for (let y = 0; y < img.length - h; y++) {
        let found = true;
        
        for (let i = 0; i < w; i++) {
          if (!found) {
            break;
          }
      
          for (let j = 0; j < h; j++) {
            if (![' ', img[y+j][x+i]].includes(monster[j][i])) {
              if (img[y+j][x+i] != 'O') {
                found = false;
              }
            }
          }
        }
        // mark monster
        if (found) {
          for (let i = 0; i < w; i++) {
            for(let j = 0; j < h; j++) {
              if (monster[j][i] != ' ') {
                img[y+j][x+i] = 'O';
              }
            }
          }
        }
      }
    }
  }

  let count = 0;
  for (const line of img) {
    for (const x of line) {
      if (x == '#') {
        count += 1;
      }
    }
  }

  return count;
}

async function main() {
	const start_time = process.hrtime();
  
  const lines = await loadInput();

  const pieces = [];

  let num = -1;
  let grid = [];
  
  for (line of lines) {
    if (line == '') {
      pieces.push(new Piece(num, grid));
      num = -1;
      grid = [];
    }
    else {
      if (num == -1) {
        num = Number(line.substring(5).replace(':',''));
      }
      else {
        grid.push(line.split(''));
      }
    }
  }
  
  pieces.push(new Piece(num, grid));

  const [part_one_ans, board] = partOne(pieces);
  const part_two_ans = partTwo(board);
  
  const diff = process.hrtime(start_time);
  const total_time = (diff[0] + diff[1]/1e9).toFixed(3);
  
  console.log(`Day 20 (${total_time}s)`);
  console.log(`  Part 1: ${part_one_ans}`);
  console.log(`  Part 2: ${part_two_ans}`);
}

if (require.main === module) {
  main();
}